### PR TITLE
chore: type check errors running `tsc`

### LINKED ISSUE #268: Type check errors running `tsc --noEmit`
### Environment

Node: v16.13
Nitropack: v0.4.4

### Reproduction

open the following Stackblitz clone repo will automatically run `tsc --noEmit` start command, then errors are logged in the `Terminal panel`.
https://stackblitz.com/edit/github-tkzhuy?file=.stackblitzrc

### Describe the bug

After running type check using typescript's `tsc --noEmit` commands, shows the following errors:

```
Errors  Files
     1  examples/cached-handler/routes/index.ts:1
     1  examples/custom-error-handler/routes/index.ts:1
     1  examples/plugins/plugins/test.ts:1
     1  src/presets/aws-lambda.ts:5
     1  src/presets/azure-functions.ts:12
     1  src/presets/azure.ts:10
     1  src/presets/firebase.ts:12
     1  src/presets/node.ts:5
     1  src/presets/stormkit.ts:5
     1  src/runtime/app.ts:43
     2  src/runtime/cache.ts:44
     1  src/runtime/client.ts:5
     1  src/runtime/entries/cloudflare.ts:30
     2  src/runtime/entries/netlify-edge.ts:16
     3  src/runtime/entries/service-worker.ts:33
     1  src/runtime/utils.ts:16
     1  test/fixture/api/kebab.ts:1
     1  test/fixture/api/param/[id].ts:1
     1  test/fixture/api/wildcard/[...param].ts:1
     1  test/fixture/routes/prerender.ts:3
     1  test/tests.ts:74
```

Errors can be separated into 3 parts:
1. **examples/ directory:**  `.nitro` folder wasn't built yet for examples in the `examples/`, which we can ignore for 'source code type checking' as end-users will build those types correctly when they are building those examples.
2. **test/ directory:** also missing  `.nitro` folder with corresponding types, which we should ignore for `test/fixture` as it can have its own `tsconfig.json` for type checks. (can be partially fix with #222)
3.  **src/  directory:** most type errors occurs, some are due to global type's name collision, some are missing discriminating union or mismatch input value.

### Additional context

_No response_

### Logs

_No response_

corresponding PR: #269

### PR DESCRIPTION
**Pull request by didavid61202**
### 🔗 Linked issue

resolves #268

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description
[ [See this PR updated code demo on stackblitz](https://stackblitz.com/edit/github-s3j1x5?file=.stackblitzrc) ]

After running type check using typescript's tsc --noEmit commands, shows some typescript type errors.

This PR aims to fix these type errors, with only type-level changes(no changes to result js), and try to have minimal type casting as possible. (some might need some future updates in upstream packages: `h3` and  `unenv`)

This PR can also fix `ci` github action check for PR #222

These type errors can be separated into 3 parts:
1. **examples/ directory:**  `.nitro` folder wasn't built yet for examples in the `examples/`, which we can ignore for 'source code type checking' as end-users will build those types correctly when they are building those examples.
2. **test/ directory:** also missing  `.nitro` folder with corresponding types, which we should ignore for `test/fixture` as it can have its own `tsconfig.json` for type checks. (can be partially fixed with #222)
3.  **src/  directory:** most type errors occur, some are due to global type's name collision, and some are missing discriminating union or mismatch input value.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by danielroe — APPROVED** [review]
Looks good to me - thank you! ❤️

File: src/runtime/entries/netlify-edge.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ export default async function (request: Request, _context) {
   })
 
   return new Response(r.body, {
-    headers: r.headers,
+    headers: r.headers as HeadersInit,
```
Comment (pi0) [inline src/runtime/entries/netlify-edge.ts:23]: I believe this was a mistake assuming the incoming request type of netlify edge was a standard Request.

Comment (pi0) [inline src/runtime/entries/netlify-edge.ts:23]: I believe this was a mistake assuming the incoming request type of netlify edge was a standard Request.

File: src/runtime/virtual/server-handlers.d.ts [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+type HandlerDefinition = {
   route: string
-  lazy?: boolean
   middleware?: boolean
-  handler: CompatibilityEventHandler | (() => Promise<CompatibilityEventHandler>)
   method?: RouterMethod
 }
+& {
+  lazy?: true
+  handler: LazyEventHandler | (() => Promise<LazyEventHandler>)
```
Comment (pi0) [inline src/runtime/virtual/server-handlers.d.ts:10]: We expect the handler still being path to a normal handler. `lazy: true`, does wrap dynamic import into lazy handler.

Comment (pi0) [inline src/runtime/virtual/server-handlers.d.ts:10]: We expect the handler still being path to a normal handler. `lazy: true`, does wrap dynamic import into lazy handler.

File: src/runtime/virtual/server-handlers.d.ts [hunk 3]
```diff
@@ -1,11 +1,17 @@
...
+}
+& {
+  lazy?: false
+  handler: CompatibilityEventHandler | (() => Promise<CompatibilityEventHandler>)
+}
```
Comment (pi0) [inline src/runtime/virtual/server-handlers.d.ts:15]: It seems reduction is not working as lazy is optional: 

image

**Review by pi0 — APPROVED** [review]

**Comment by pi0** [comment]
Thanks for working on this @didavid61202 @danielroe <3 I've made few modifications to make it mergeable as soon as possible. Feel free to open a subsequent PR to improve.

File: src/runtime/virtual/server-handlers.d.ts (hunk 3 above)
Comment (pi0) [reply src/runtime/virtual/server-handlers.d.ts:12]: (updated: added back as a more compact type)

**Pull request merged**